### PR TITLE
Python Refactor of Local Scripts

### LINKED ISSUE #3: Port cluster-management scripts to Python
Motivations:
* greater flexibility for users to tweak and override
* use a config file format (YAML?) that will not be executed (dangerous)
* a decent language/framework for testing that is the same as the project's main language
* language popularity / developer base

At least convert the scripts that run locally and consume a config file. The scripts that run on EMR can be left alone, if preferred.

### PR DESCRIPTION
**Pull request by nelsonw2014**
Attempt to resolve #3.

### CODE REVIEW

File: watershed.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        'forward-local-ports',
+        help="Forward the local ports for ssh access."
+    )
+    forward_local_ports_parser.set_defaults(which="forward-local-ports")
+    forward_local_ports_parser.add_argument(*_cluster_id_args, **_cluster_id_kwargs)
+    forward_local_ports_parser.add_argument(*_private_key_args, **_private_key_kwargs)
+    forward_local_ports_parser.add_argument(*_profile_args, **_profile_kwargs)
+    terminate_clusters_parser = subparsers.add_parser(
+        'terminate-clusters',
+        help="Terminate the cluster(s) that are associated with the provided luster ID(s)"
```
Comment (pmogren) [inline watershed.py]: typo 'luster'

File: watershed.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        required=True,
+        type=argparse.FileType('r')
+    )
+    _cluster_id_args = [
+        "-i",
+        "--cluster-id"
+    ]
+    _cluster_id_kwargs = dict(
+        action="store",
+        help="Specify the cluster id for the operation to be preformed on",
```
Comment (pmogren) [inline watershed.py]: typo 'preformed'

File: watershed.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+import argparse
+import json
+from aws_tools.s3 import upload_resources
+from aws_tools.emr import launch_emr_cluster, terminate_emr_cluster
+from ssh_tools.ssh import forward_necessary_ports
+
+
+def parse_arguments():
+    _fc_description = """
+Python/Boto solution which compliments kinesis with:
```
Comment (pmogren) [inline watershed.py]: s/kinesis/Amazon Kinesis/

File: conf/defaults.json [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      "resourcesBucket": "emr-resources",
+      "resourcesDirectory": "resources"
+    },
+    "EMR": {
+      "clusterName": "stream-archival",
+      "roles": {
+        "service": "EMR-Role-Name",
+        "ec2": "EMR-EC2-Resource-Role-Name"
+      },
+      "ec2KeyName": "key-pair-name",
```
Comment (pmogren) [inline conf/defaults.json]: ec2KeyName, ec2SubnetId, and additionalMasterSecurityGroups have defaults that aren't really valid as defaults. Is it necessary to put something here?

File: aws_tools/emr.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                    '-hiveconf',
+                    'tablename='+emr_config['hive']['tableName'],
+                    '-hiveconf',
+                    'streamname='+emr_config['kinesis']['streamName']
+                ]
+            },
+            'ActionOnFailure': 'CANCEL_AND_WAIT'
+        },
+        {
+            'Name': 'Install Hive',
```
Comment (pmogren) [inline aws_tools/emr.py]: Swap the order of these, install Hive before creating a table in it

File: watershed.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        help="Launch a cluster."
+    )
+    launch_cluster_parser.set_defaults(which="launch-cluster")
+    launch_cluster_parser.add_argument(*_config_file_args, **_config_file_kwargs)
+    launch_cluster_parser.add_argument(
+        '-w',
+        '--wait-until-ready',
+        action="store_const",
+        const=True,
+        help="Wait until cluster launches (takes up to 10 minutes at most)"
```
Comment (pmogren) [inline watershed.py]: Clusters that are large, spot-priced, or misconfigured can take more than 10 minutes to start. Why not stick with the "usually takes five minutes or more" advice?

**Comment by pmogren** [comment]
Overall this looks really good and is much appreciated. I think it sacrificed some of the highly flexible parts of the configuration, but we can iterate on restoring as needed. Thanks!

File: aws_tools/s3.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+
+def upload_resources(s3_config=None, profile="default", force_upload=False):
+    s3_resource = boto3.session.Session(profile_name=profile).resource('s3')
+
+    try:
+        file_upload_count = 0
+        project_root = os.path.dirname(os.path.abspath(__file__)).replace('/aws_tools', '/')
+        if s3_resource.Bucket(s3_config['resourcesBucket']) in s3_resource.buckets.all() and not force_upload:
+            raise ValueError("Bucket already exists, run with -F to force upload.")
```
Comment (pmogren) [inline aws_tools/s3.py]: Please replace -F with -f throughout.

File: aws_tools/s3.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        s3_resource.create_bucket(Bucket=s3_config['resourcesBucket'])
+        for folder in os.walk(project_root+'resources/s3'):
+            # for all directories within the local_resources_directory
+            if folder[2]:
+                # if there are files in the directory
+                for file in folder[2]:
+                    # for every file in the directory
+                    s3_file_path = s3_config['resourcesDirectory']+folder[0].split('resources/s3')[1]+"/"+file
```
Comment (pmogren) [inline aws_tools/s3.py]: The usual S3 terminology I've seen for a full "directory" path is officially "prefix", colloquially "path"; the individual parts thereof "folder".  Let's not use "directory".

File: aws_tools/emr.py (hunk 5 above)
Comment (nelsonw2014) [inline aws_tools/emr.py]: the list order goes backwards. I have no idea why, that's just what testing has revealed to me

File: conf/defaults.json (hunk 4 above)
Comment (nelsonw2014) [inline conf/defaults.json]: They can be empty strings and lists if you want. I wanted the default to have a good representation of content expected. There should probably be some sort of better documentation of what the configuration file entails, but I'm not sure how to go about that best.

**Comment by pmogren** [comment]
When watershed.py is run with zero arguments, it should print the nice help message rather than a stack trace, but still exit with a non-zero code.

File: conf/defaults.json (hunk 4 above)
Comment (pmogren) [inline conf/defaults.json]: Defaults for EMR.roles.* should be the names that AWS generates, as they were in the Bash config

File: watershed.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
+# Copyright (C) 2015 Commerce Technologies, Inc.
```
Comment (pmogren) [inline watershed.py]: Add shebang (first line should read #!/usr/bin/env python3) and chmod this file to be executable.

File: watershed.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        "--profile"
+    ]
+    _profile_kwargs = dict(
+        action="store",
+        help="Specify the AWS profile to use. Defaults to 'default'",
+        default='default'
+    )
+
+    parser = argparse.ArgumentParser(
+        prog="Watershed",
```
Comment (pmogren) [inline watershed.py]: This prints as the command name in usage message, so gives bad information. "watershed.py" would suffice, but I'd like to have the equivalent to $0 in Bash. http://stackoverflow.com/questions/4152963/get-the-name-of-current-script-with-python

**Comment by pmogren** [comment]
I did some manual testing: create cluster, forward ports, query stream, query stream archive, terminate cluster.  Ready to merge after the tweaks requested today are resolved.

**Comment by pmogren** [comment]
After testing, I have the following directories in my repo. Please add to .gitignore (using a wildcard for the first part)

	aws_tools/__pycache__/
	ssh_tools/__pycache__/

**Pull request merged**